File: brute.cpp
#include <iostream>
using namespace std;
int R, C, N;
int Max_contrasts = -1;
bool wall[100][100], best_wall[100][100];

void print_wall(bool w[][100]) {
  int i, j;
  for (i = 0; i < R; i++) {
    for (j = 0; j < C; j++) {
      cout << w[i][j];
    }
    cout << endl;
  }
}

// count the number of contrasts in the boolean 2D array wall
int count_contrasts() {
	int i, j, ans = 0;
  for (i = 0; i < R; i++) {
    for (j = 0; j < C; j++) {
      if ((j != C-1) && (wall[i][j] != wall[i][j+1])) {
        ans++;
      }

      if ((i != R-1) && (wall[i][j] != wall[i+1][j])) {
        ans++;
      }
    }
  }
	return ans;
}

void combinations(int k, int offset) {
	int i, j;
	if (k==0) {
		int n_contrasts = count_contrasts();
		if (n_contrasts > Max_contrasts) {
			Max_contrasts = n_contrasts;
      for (i=0; i<R; i++)
        for (j=0; j<C; j++)
          best_wall[i][j] = wall[i][j];
		}
		return;
	}
	for (i=offset; i<R*C; i++) {
    wall[i/C][i%C] = true;
		combinations(k-1, i+1);
    wall[i/C][i%C] = false;
	}
}

int main() {
  cin >> R >> C >> N;
  combinations(N, 0);
  cerr << Max_contrasts << endl;
  print_wall(best_wall);
  return 0;
}

File: brute.py
from itertools import combinations as comb
from itertools import product


def paint(black_cells):
    wall = [[False] * C for _ in range(R)]
    for r, c in black_cells:
        wall[r][c] = True
    return wall


def count_contrasts(black_cells):
    wall = paint(black_cells)
    count = 0
    for r, c in black_cells:
        if r > 0:
            count += (wall[r-1][c] != wall[r][c])
        if r < R-1:
            count += (wall[r+1][c] != wall[r][c])
        if c > 0:
            count += (wall[r][c-1] != wall[r][c])
        if c < C-1:
            count += (wall[r][c+1] != wall[r][c])
    return count


R, C, N = map(int, input().split())
grid = product(range(R), range(C))
max_contrasts = -1
for cells in comb(grid, N):
    count = count_contrasts(cells)
    if count > max_contrasts:
        max_contrasts = count
        best_wall = paint(cells)
for row in best_wall:
    for j in row:
        print(int(j), end='')
    print()
